File: lib/frontmatter.py
"""Copy a subtree, filling missing frontmatter keys in its markdown files.

A borrowed harness speaks the source client's dialect, and a receiving client
may require a frontmatter field the source never writes: agens refuses an
agent definition without `mode:`, while Claude Code has no such field at all.
The receiving client's requirement belongs to the projection, not to the
source files, so the key is inserted while the copy is made.

Only a missing key is filled. A file that already states the key keeps its own
value, whatever it is, because a default that overwrote a stated value would
be a rename wearing a default's name. Files without a frontmatter block are
copied untouched: inventing a block would also invent the required fields the
receiving client checks before this one, and a loud parse failure there beats
a silently half-valid definition here.

Only markdown files are considered, and only text is touched. Anything that is
not valid UTF-8 is copied byte for byte.
"""

from __future__ import annotations

import argparse
import os
import re
import shutil
import sys
from pathlib import Path


def fill_frontmatter(text: str, defaults: dict[str, str]) -> str:
    """Insert each missing top-level key just before the closing delimiter."""
    lines = text.split("\n")
    if not lines or lines[0].strip() != "---":
        return text

    closing = None
    for index in range(1, len(lines)):
        if lines[index].strip() == "---":
            closing = index
            break
    if closing is None:
        return text

    stated = {
        match.group(1)
        for line in lines[1:closing]
        if (match := re.match(r"([A-Za-z0-9_-]+)\s*:", line))
    }

    missing = [key for key in sorted(defaults) if key not in stated]
    if not missing:
        return text

    inserted = [f"{key}: {defaults[key]}" for key in missing]

    return "\n".join(lines[:closing] + inserted + lines[closing:])


def fill(data: bytes, defaults: dict[str, str]) -> bytes:
    try:
        text = data.decode("utf-8")
    except UnicodeDecodeError:
        return data

    return fill_frontmatter(text, defaults).encode("utf-8")


def copy_tree(source: Path, target: Path, defaults: dict[str, str]) -> None:
    """Copy through symbolic links, since the source is a store tree of them."""
    if source.is_dir():
        target.mkdir(parents=True, exist_ok=True)
        for entry in sorted(os.listdir(source)):
            copy_tree(source / entry, target / entry, defaults)
        return

    target.parent.mkdir(parents=True, exist_ok=True)
    data = source.read_bytes()
    if source.suffix == ".md":
        data = fill(data, defaults)
    target.write_bytes(data)
    shutil.copymode(source, target)
    target.chmod(target.stat().st_mode | 0o200)


def main() -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--source", required=True, type=Path, help="subtree to copy")
    parser.add_argument("--target", required=True, type=Path, help="where the filled copy is written")
    parser.add_argument(
        "--default",
        action="append",
        default=[],
        metavar="KEY=VALUE",
        help="frontmatter key filled into markdown files that lack it; repeatable",
    )
    arguments = parser.parse_args()

    defaults = {}
    for pair in arguments.default:
        key, separator, value = pair.partition("=")
        if not separator or not key:
            print(f"gentle-ai: --default expects KEY=VALUE, got {pair!r}", file=sys.stderr)
            return 1
        defaults[key] = value

    copy_tree(arguments.source, arguments.target, defaults)

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: internal/frontmatter/frontmatter.go
// Package frontmatter ports lib/frontmatter.py: copying a subtree while
// filling missing frontmatter keys in its markdown files.
//
// Only a missing key is filled -- a file that already states the key
// keeps its own value, whatever it is, because a default that overwrote a
// stated value would be a rename wearing a default's name. Files without
// a frontmatter block, or with an unclosed one, are copied untouched:
// inventing a block would also invent the required fields the receiving
// client checks before this one, and a loud parse failure there beats a
// silently half-valid definition here.
package frontmatter

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"unicode/utf8"
)

var frontmatterKeyPattern = regexp.MustCompile(`^([A-Za-z0-9_-]+)\s*:`)

// FillFrontmatter is fill_frontmatter from frontmatter.py: insert each
// default key defaults names but text does not already state, just
// before the closing "---" delimiter, in sorted key order. Text without
// an opening "---" on its first line, or without a matching closing
// "---" line, is returned unchanged.
func FillFrontmatter(text string, defaults map[string]string) string {
	lines := strings.Split(text, "\n")
	if len(lines) == 0 || strings.TrimSpace(lines[0]) != "---" {
		return text
	}

	closing := -1
	for i := 1; i < len(lines); i++ {
		if strings.TrimSpace(lines[i]) == "---" {
			closing = i
			break
		}
	}
	if closing == -1 {
		return text
	}

	stated := make(map[string]bool)
	for _, line := range lines[1:closing] {
		if match := frontmatterKeyPattern.FindStringSubmatch(line); match != nil {
			stated[match[1]] = true
		}
	}

	keys := make([]string, 0, len(defaults))
	for key := range defaults {
		keys = append(keys, key)
	}
	sort.Strings(keys)

	var missing []string
	for _, key := range keys {
		if !stated[key] {
			missing = append(missing, key)
		}
	}
	if len(missing) == 0 {
		return text
	}

	inserted := make([]string, len(missing))
	for i, key := range missing {
		inserted[i] = fmt.Sprintf("%s: %s", key, defaults[key])
	}

	result := make([]string, 0, len(lines)+len(inserted))
	result = append(result, lines[:closing]...)
	result = append(result, inserted...)
	result = append(result, lines[closing:]...)

	return strings.Join(result, "\n")
}

// Fill is fill() from frontmatter.py: apply FillFrontmatter to data
// decoded as UTF-8, or return data unchanged when it does not decode.
func Fill(data []byte, defaults map[string]string) []byte {
	if !utf8.Valid(data) {
		return data
	}
	return []byte(FillFrontmatter(string(data), defaults))
}

// CopyTree is copy_tree from frontmatter.py: recurse through source,
// mirroring its structure exactly under target, filling frontmatter only
// in files named "*.md" and copying through symbolic links since the
// source is a store tree of them. Every delivered file gains the
// owner-write bit, since the store copy it came from is read-only.
func CopyTree(source, target string, defaults map[string]string) error {
	info, err := os.Stat(source)
	if err != nil {
		return err
	}

	if info.IsDir() {
		if err := os.MkdirAll(target, 0o777); err != nil {
			return err
		}
		entries, err := os.ReadDir(source)
		if err != nil {
			return err
		}
		names := make([]string, len(entries))
		for i, entry := range entries {
			names[i] = entry.Name()
		}
		sort.Strings(names)

		for _, name := range names {
			if err := CopyTree(filepath.Join(source, name), filepath.Join(target, name), defaults); err != nil {
				return err
			}
		}
		return nil
	}

	if err := os.MkdirAll(filepath.Dir(target), 0o777); err != nil {
		return err
	}

	data, err := os.ReadFile(source)
	if err != nil {
		return err
	}
	if filepath.Ext(source) == ".md" {
		data = Fill(data, defaults)
	}
	if err := os.WriteFile(target, data, info.Mode().Perm()); err != nil {
		return err
	}

	return os.Chmod(target, info.Mode().Perm()|0o200)
}

// ParseDefaultArgument splits a --default KEY=VALUE argument the way
// main()'s partition("=") does: the first "=" separates the two halves,
// and either a missing "=" or an empty KEY is rejected.
func ParseDefaultArgument(arg string) (key, value string, err error) {
	k, v, found := strings.Cut(arg, "=")
	if !found || k == "" {
		return "", "", fmt.Errorf("--default expects KEY=VALUE, got %q", arg)
	}
	return k, v, nil
}
